File: src/main/java/net/cjsah/console/command/builder/ArgumentBuilder.java
package net.cjsah.console.command.builder;

import net.cjsah.console.command.Command;
import net.cjsah.console.command.source.CommandSource;
import net.cjsah.console.command.tree.CommandNode;
import net.cjsah.console.command.tree.RootCommandNode;

import java.util.Collection;
import java.util.function.Predicate;

public abstract class ArgumentBuilder<T extends ArgumentBuilder<T>> {
    private final RootCommandNode arguments = new RootCommandNode();
    private Predicate<CommandSource<?>> requirement = (source) -> true;
    private Command command = null;
    private String help = "";

    public T then(ArgumentBuilder<?> argument) {
        this.arguments.addChild(argument.build());
        return getThis();
    }

    public T executes(String help, Command command) {
        this.command = command;
        this.help = help;
        return getThis();
    }

    public T requires(Predicate<CommandSource<?>> requirement) {
        this.requirement = requirement;
        return getThis();
    }

    protected abstract T getThis();

    public abstract CommandNode build();

    public Collection<CommandNode> getArguments() {
        return arguments.getChildren();
    }

    public Predicate<CommandSource<?>> getRequirement() {
        return this.requirement;
    }

    public Command getCommand() {
        return this.command;
    }

    public String getHelp() {
        return this.help;
    }
}

File: src/main/kotlin/net/cjsah/console/command/builder/ArgumentBuilder.kt
package net.cjsah.console.command.builder

import net.cjsah.console.command.Command
import net.cjsah.console.command.source.CommandSource
import net.cjsah.console.command.tree.CommandNode
import net.cjsah.console.command.tree.RootCommandNode
import java.util.function.Predicate

abstract class ArgumentBuilder<T : ArgumentBuilder<T>> {
    private val arguments = RootCommandNode()
    private var requirement: Predicate<CommandSource<*>> = Predicate { true }
    private var command: Command? = null
    private var help = ""

    open fun then(argument: ArgumentBuilder<*>): T {
        arguments.addChild(argument.build())
        return getThis()
    }

    open fun executes(help: String, command: Command?): T {
        this.command = command
        this.help = help
        return getThis()
    }

    open fun requires(requirement: Predicate<CommandSource<*>>): T {
        this.requirement = requirement
        return getThis()
    }

    protected abstract fun getThis(): T

    abstract fun build(): CommandNode

    open fun getArguments(): Collection<CommandNode> {
        return arguments.getChildren()
    }

    open fun getRequirement(): Predicate<CommandSource<*>> {
        return requirement
    }

    open fun getCommand(): Command? {
        return command
    }

    open fun getHelp(): String {
        return help
    }

}